File: evo/001-gen_word.js
'use strict'

function choose(arr) {
  return arr[Math.floor(Math.random() * arr.length)]
}

function gen(letters /* string[] */, word /* string[] */, attempt /* string[] */) {
  if (attempt.length === word.length) {
    let i = 0
    while (i < attempt.length && attempt[i] === word[i]) {
      i++
    }
    return i === attempt.length
  }
  attempt.push(choose(letters))
  return gen(letters, word, attempt)
}

const letters = 'abcdefghijklmnopqrstuvwxyz'.split('')
const word = 'paste'.split('')
let found = false
let attempts = 0

function setup() {
  performance.mark('start')
  createCanvas(460, 200)
  textSize(16)
  console.log('start')
  noLoop()
}

function draw() {
  background(0)
  fill(255)
  while (!found) {
    found = gen(letters, word, [])
    attempts++
  }

  console.log('end')
  performance.mark('end')
  const m = performance.measure('run', 'start', 'end')
  const { duration } = m

  text(
    `found '${word.join('')}' in ${attempts} attempts, ${duration.toFixed(2)}ms, ${(attempts / duration).toFixed(
      2
    )} a/ms`,
    10,
    20
  )
}

File: evo/001-gen_word.py
import random
import time


def gen(
    letters: list[str], word: list[str], attempt: list[str], depth: int, word_len: int
) -> bool:
    if depth == word_len:
        i = 0
        while i < word_len and attempt[i] == word[i]:
            i += 1
        return i == word_len
    else:
        attempt[depth] = random.choice(letters)
        return gen(letters, word, attempt, depth + 1, word_len)


if __name__ == "__main__":
    start = time.time()
    letters = list("abcdefghijklmnopqrstuvwxyz")
    word = list("paste")
    word_len = len(word)
    found = False
    attempts = 0
    try:
        while not found:
            found = gen(letters, word, [None for i in range(word_len)], 0, word_len)
            attempts += 1
            if attempts % 100_000 == 0:
                print(".", end="", flush=True)
        print(f"found '{''.join(word)}' in {attempts} attempts,", end=" ")
    except KeyboardInterrupt:
        print(f"exiting after {attempts} attempts,", end=" ")

    finish = time.time()
    ms = (finish - start) * 1000
    print(f"took {ms:.4f}ms, {attempts / ms:.2f} a/ms")
